### ISSUE TITLE
simple resnet18 outputs different between pytorch & tensorrt

### ISSUE THREAD

**Issue by ucasqcz** [NVIDIA/TensorRT#3056]
```
import tensorrt as trt

import torch

TRT_LOGGER = trt.Logger(trt.Logger.VERBOSE)
EXPLICIT_BATCH = 1 << int(trt.NetworkDefinitionCreationFlag.EXPLICIT_BATCH)
def torch_dtype_from_trt(dtype: trt.DataType) -> torch.dtype:
    """Convert pytorch dtype to TensorRT dtype.

    Args:
        dtype (str.DataType): The data type in tensorrt.

    Returns:
        torch.dtype: The corresponding data type in torch.
    """

    if dtype == trt.bool:
        return torch.bool
    elif dtype == trt.int8:
        return torch.int8
    elif dtype == trt.int32:
        return torch.int32
    elif dtype == trt.float16:
        return torch.float16
    elif dtype == trt.float32:
        return torch.float32
    else:
        raise TypeError(f'{dtype} is not supported by torch')


class TRTWrapper(torch.nn.Module):

    def __init__(self, engine, output_names):
        super().__init__()
        self.engine = engine
        if isinstance(self.engine, str):
            with trt.Logger() as logger, trt.Runtime(logger) as runtime:
                with open(self.engine, mode='rb') as f:
                    engine_bytes = f.read()
                self.engine = runtime.deserialize_cuda_engine(engine_bytes)
        self.context = self.engine.create_execution_context()
        names = [_ for _ in self.engine]
        input_names = list(filter(self.engine.binding_is_input, names))
        self._input_names = input_names
        self._output_names = output_names

        if self._output_names is None:
            output_names = list(set(names) - set(input_names))
            self._output_names = output_names

    def forward(self, inputs):
        bindings = [None] * (len(self._input_names) + len(self._output_names))
        for input_name, input_tensor in inputs.items():
            idx = self.engine.get_binding_index(input_name)
            self.context.set_binding_shape(idx, tuple(input_tensor.shape))
            bindings[idx] = input_tensor.contiguous().data_ptr()

        outputs = {}
        for output_name in self._output_names:
            idx = self.engine.get_binding_index(output_name)
            dtype = torch_dtype_from_trt(self.engine.get_binding_dtype(idx))
            shape = tuple(self.context.get_binding_shape(idx))

            device = torch.device('cuda')
            output = torch.zeros(size=shape, dtype=dtype, device=device)
            outputs[output_name] = output
            bindings[idx] = output.data_ptr()
        self.context.execute_async_v2(bindings, torch.cuda.current_stream().cuda_stream)
        return outputs
def toy():
    from torchvision.models.resnet import resnet18
    torch_model = resnet18(pretrained=True).eval().cuda()

    inputs = torch.rand([1, 3, 224, 224]).cuda().contiguous()
    with torch.no_grad():
        torch.onnx.export(
            torch_model,
            inputs,
            'toy.onnx',
            input_names=['input'],
            output_names=['output'],
            opset_version=17,
        )
    max_workspace_size = 1
    builder = trt.Builder(TRT_LOGGER)
    network = builder.create_network(EXPLICIT_BATCH)
    config = builder.create_builder_config()
    parser = trt.OnnxParser(network, TRT_LOGGER)
    config.set_memory_pool_limit(trt.MemoryPoolType.WORKSPACE, (1 << 32) * max_workspace_size)
    with open('toy.onnx', 'rb') as model:
        if not parser.parse(model.read()):
            print('ERROR: Failed to parse the onnx file')
            for error in range(parser.num_errors):
                print(parser.get_error(error))
            return None
    plan = builder.build_serialized_network(network, config)

    print('Completed creating engine')
    with open('toy.trt', "wb") as f:
        f.write(plan)
    trt_model = TRTWrapper('toy.trt', None)

    trt_outputs = trt_model({'input': inputs})

    torch_outputs = torch_model(inputs)
    diff_torch2trt = torch_outputs - trt_outputs['output']

    print(f'diff_torch2trt max: {diff_torch2trt.max()}, mean: {diff_torch2trt.abs().mean()}')
if __name__ == '__main__':
    toy()

# diff_torch2trt max: 0.005703568458557129, mean: 0.0012052928796038032
```

**Comment by ucasqcz** [comment]
the output of torch and tensorrt for base model like res18 should be same

**Comment by ucasqcz** [comment]
> polygraphy run toy.onnx --trt --onnxrt
```
[I] Building engine with configuration:
    Flags                  | []
    Engine Capability      | EngineCapability.DEFAULT
    Memory Pools           | [WORKSPACE: 81085.31 MiB, TACTIC_DRAM: 0.00 MiB]
    Tactic Sources         | [CUBLAS, CUBLAS_LT, CUDNN, EDGE_MASK_CONVOLUTIONS, JIT_CONVOLUTIONS]
    Profiling Verbosity    | ProfilingVerbosity.DETAILED
[I] Finished engine building in 12.280 seconds
[I] trt-runner-N0-06/10/23-09:51:10
    ---- Inference Input(s) ----
    {input [dtype=float32, shape=(1, 3, 224, 224)]}
[I] trt-runner-N0-06/10/23-09:51:10
    ---- Inference Output(s) ----
    {output [dtype=float32, shape=(1, 1000)]}
[I] trt-runner-N0-06/10/23-09:51:10     | Completed 1 iteration(s) in 2.276 ms | Average inference time: 2.276 ms.
[I] onnxrt-runner-N0-06/10/23-09:51:10  | Activating and starting inference
[I] Creating ONNX-Runtime Inference Session with providers: ['CPUExecutionProvider']
[I] onnxrt-runner-N0-06/10/23-09:51:10
    ---- Inference Input(s) ----
    {input [dtype=float32, shape=(1, 3, 224, 224)]}
[I] onnxrt-runner-N0-06/10/23-09:51:10
    ---- Inference Output(s) ----
    {output [dtype=float32, shape=(1, 1000)]}
[I] onnxrt-runner-N0-06/10/23-09:51:10  | Completed 1 iteration(s) in 7.67 ms | Average inference time: 7.67 ms.
[I] Accuracy Comparison | trt-runner-N0-06/10/23-09:51:10 vs. onnxrt-runner-N0-06/10/23-09:51:10
[I]     Comparing Output: 'output' (dtype=float32, shape=(1, 1000)) with 'output' (dtype=float32, shape=(1, 1000))
[I]         Tolerance: [abs=1e-05, rel=1e-05] | Checking elemwise error
[I]         trt-runner-N0-06/10/23-09:51:10: output | Stats: mean=8.1329e-06, std-dev=1.4193, var=2.0144, median=-0.096946, min=-4.1115 at (0, 865), max=5.2154 at (0, 111), avg-magnitude=1.1274
[I]         onnxrt-runner-N0-06/10/23-09:51:10: output | Stats: mean=8.1329e-06, std-dev=1.4193, var=2.0144, median=-0.096946, min=-4.1115 at (0, 865), max=5.2154 at (0, 111), avg-magnitude=1.1274
[I]         Error Metrics: output
[I]             Minimum Required Tolerance: elemwise error | [abs=3.3379e-06] OR [rel=0.0016508] (requirements may be lower if both abs/rel tolerances are set)
[I]             Absolute Difference | Stats: mean=6.6539e-07, std-dev=4.9263e-07, var=2.4268e-13, median=5.9605e-07, min=0 at (0, 63), max=3.3379e-06 at (0, 644), avg-magnitude=6.6539e-07
[I]             Relative Difference | Stats: mean=4.0166e-06, std-dev=5.398e-05, var=2.9139e-09, median=6.1401e-07, min=0 at (0, 63), max=0.0016508 at (0, 217), avg-magnitude=4.0166e-06
[I]         PASSED | Output: 'output' | Difference is within tolerance (rel=1e-05, abs=1e-05)
[I]     PASSED | All outputs matched | Outputs: ['output']
[I] Accuracy Summary | trt-runner-N0-06/10/23-09:51:10 vs. onnxrt-runner-N0-06/10/23-09:51:10 | Passed: 1/1 iterations | Pass Rate: 100.0%
[I] PASSED | Runtime: 18.352s | Command: /usr/local/bin/polygraphy run toy.onnx --trt --onnxrt
```
the diff passes the tolerance, so what's the different between polygraphy and mycode ? do you have any suggestion for debug this ?

**Comment by ucasqcz** [comment]
meanwhile, is the below normal?

**Comment by zerollzeng** [comment]
So we can know the diff between trt and onnxrt is good. Usually it's caused by inputs/synchronization. or the problem is in onnx export(unlikely in your case)

**Comment by zerollzeng** [comment]
> meanwhile, is the below normal? the onnxruntime runs with 'CPUExecutionProvider'

yes it's normal.

**Comment by ucasqcz** [comment]
> it's caused by inputs/synchronization

what does the inputs

**Comment by ucasqcz** [comment]
we have modify the code using TRTModule instead of TRTWarpper
```
import torch
import tensorrt as trt
from torch_tensorrt.fx import TRTModule


TRT_LOGGER = trt.Logger(trt.Logger.VERBOSE)
EXPLICIT_BATCH = 1 << int(trt.NetworkDefinitionCreationFlag.EXPLICIT_BATCH)

def toy():
    from torchvision.models.resnet import resnet18
    torch_model = resnet18(pretrained=True).eval().cuda()

    inputs = torch.rand([1, 3, 224, 224]).cuda().contiguous()
    with torch.no_grad():
        torch.onnx.export(
            torch_model,
            inputs,
            'toy.onnx',
            input_names=['input'],
            output_names=['output'],
            opset_version=17,
        )
    max_workspace_size = 1
    builder = trt.Builder(TRT_LOGGER)
    network = builder.create_network(EXPLICIT_BATCH)
    config = builder.create_builder_config()
    parser = trt.OnnxParser(network, TRT_LOGGER)
    config.set_memory_pool_limit(trt.MemoryPoolType.WORKSPACE, (1 << 32) * max_workspace_size)
    with open('toy.onnx', 'rb') as model:
        if not parser.parse(model.read()):
            print('ERROR: Failed to parse the onnx file')
            for error in range(parser.num_errors):
                print(parser.get_error(error))
            return None
    plan = builder.build_serialized_network(network, config)

    print('Completed creating engine')
    with open('toy.trt', "wb") as f:
        f.write(plan)

    with trt.Logger() as logger, trt.Runtime(logger) as runtime:
        with open('toy.trt', mode='rb') as f:
            engine_bytes = f.read()
        engine = runtime.deserialize_cuda_engine(engine_bytes)

    trt_model = TRTModule(engine, input_names=['input'], output_names=['output'])

    trt_outputs = trt_model(*[inputs])

    torch_outputs = torch_model(inputs)
    diff_torch2trt = torch_outputs - trt_outputs

    print(f'diff_torch2trt max: {diff_torch2trt.max()}, mean: {diff_torch2trt.abs().mean()}')

if __name__ == '__main__':
    toy()

# diff_torch2trt max: 0.006762504577636719, mean: 0.0014320116024464369
```
still large error

**Comment by zerollzeng** [comment]
> what does the inputs / sync mean?

make sure to feed the same inputs and do a cuda sync after inference so the result is transferred to host memory correctly.

**Comment by ucasqcz** [comment]
> > what does the inputs / sync mean?
> 
> make sure to feed the same inputs and do a cuda sync after inference so the result is transferred to host memory correctly.

as the above code, we feed the trt_model & torch_model with the same cuda tensor, meanwhile, we diff  the output cuda tensor, no need to device2host copy. can you reproduce the large diff on your env?

**Comment by zerollzeng** [comment]
> self.context.execute_async_v2(bindings, torch.cuda.current_stream().cuda_stream)

Could you please try use execute_v2() here? see https://docs.nvidia.com/deeplearning/tensorrt/api/python_api/infer/Core/ExecutionContext.html

**Comment by ucasqcz** [comment]
> > self.context.execute_async_v2(bindings, torch.cuda.current_stream().cuda_stream)
> 
> Could you please try use execute_v2() here? see https://docs.nvidia.com/deeplearning/tensorrt/api/python_api/infer/Core/ExecutionContext.html

nothing changed when i use self.context.execute_v2(bindings)

**Comment by zerollzeng** [comment]
Okay... I think we can check the diff between pytorch and onnx first. we see the diff between onnxruntime and trt is very small, maybe it's caused by export.

**Comment by ttyio** [comment]
Usually we did not use random input for accuracy debug, they are mostly floating point arithmetic error.

Pls run the model on benchmark to debug the accuracy issue, thanks!
